libcloud/compute/drivers/scaleway.py

```diff
@@ -0,0 +1,665 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Scaleway Driver
+"""
+
+import copy
+try:
+    import simplejson as json
+except ImportError:
+    import json
+
+from libcloud.common.base import ConnectionUserAndKey, JsonResponse
+from libcloud.common.types import ProviderError
+from libcloud.compute.base import NodeDriver, NodeImage, Node, NodeSize
+from libcloud.compute.base import NodeLocation
+from libcloud.compute.base import StorageVolume, VolumeSnapshot, KeyPair
+from libcloud.compute.providers import Provider
+from libcloud.compute.types import NodeState, VolumeSnapshotState
+from libcloud.utils.iso8601 import parse_date
+from libcloud.utils.py3 import httplib
+
+__all__ = [
+    'ScalewayResponse',
+    'ScalewayConnection',
+    'ScalewayNodeDriver'
+]
+
+SCALEWAY_API_HOSTS = {
+    'default': 'api.scaleway.com',
+    'account': 'account.scaleway.com',
+    'par1': 'cp-par1.scaleway.com',
+    'ams1': 'cp-ams1.scaleway.com',
+}
+
+# The API doesn't give location info, so we provide it ourselves, instead.
+SCALEWAY_LOCATION_DATA = [
+    {'id': 'par1', 'name': 'Paris 1', 'country': 'FR'},
+    {'id': 'ams1', 'name': 'Amsterdam 1', 'country': 'NL'},
+]
+
+
+class ScalewayResponse(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED,
+                            httplib.CREATED, httplib.NO_CONTENT]
+
+    def parse_body(self):
+        body = super(ScalewayResponse, self).parse_body()
+
+        links = self.connection.connection.getresponse().links
+        if links and 'next' in links:
+            response = self.connection.request(links['next']['url'],
+                                               data=self.connection.data,
+                                               method=self.connection.method)
```

Comment:
This is quite surprising: a connection request while parsing the body? Did you have another option? Also it looks it only fetches the next link while we could have more urls to fetch?

Reply:
It recurses, yes. It's impossible to know how many pages will be in a collection until the initial request is made, so multi-page results need to be processed afterward, but there aren't a lot of points where these further requests could be made without having to place the pagination logic outside the response itself, which seems to miss the purpose of putting responses in their own class to begin with. So it's done in the body parsing section, which is actually a bonus anyway, as the response will then always contain all the pages in a given collection, with the first page appending the recursive result of the next to its own. (And, given the nature of recursion, the second appends the result of the third to its own, so that each level of recursion returns its own results plus any others until the end of the list is reached.)

I'm open to other suggestions, if you have any insights into the design that I've missed. (Which seems likely.)

Reply:
I'll look at the solutions adopter in other drivers. (Another solution that I don't like is GCE which makes pagination explicit.)

Reply:
I think the ECS approach is better, which makes pagination explicit. Do you think it would be easy to do? If it's more than say one hour or two of work, it's probably not worth it as the current approach does work even if it mixes concerns and is very surprising.